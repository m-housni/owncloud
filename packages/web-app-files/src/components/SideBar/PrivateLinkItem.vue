<template>
  <oc-button
    v-oc-tooltip="buttonLabel"
    appearance="raw"
    :aria-label="buttonLabel"
    class="oc-files-private-link-copy-url"
    :variation="copied ? 'success' : 'passive'"
    @click="copyPrivateLinkToClipboard"
  >
    <span v-text="buttonText" />
    <oc-icon
      v-if="copied"
      key="oc-copy-to-clipboard-copied"
      name="checkbox-circle"
      class="_clipboard-success-animation"
    />
    <oc-icon v-else key="oc-copy-to-clipboard-copy" name="clipboard" />
  </oc-button>
</template>

<script>
import { mapActions } from 'vuex'
import copyToClipboard from 'copy-to-clipboard'
import { unref } from 'vue'

export default {
  name: 'PrivateLinkItem',
  inject: ['displayedItem'],
  data: () => ({
    copied: false,
    timeout: null
  }),
  computed: {
    buttonText() {
      return this.$gettext('Private link')
    },
    buttonLabel() {
      return this.$gettext('Copy private link to clipboard')
    }
  },
  methods: {
    ...mapActions(['showMessage']),
    copyPrivateLinkToClipboard() {
      copyToClipboard(unref(this.displayedItem).privateLink)
      this.clipboardSuccessHandler()
      this.showMessage({
        title: this.$gettext('Private link copied'),
        desc: this.$gettext('The private link has been copied to your clipboard.')
      })
    },
    clipboardSuccessHandler() {
      this.copied = true
      clearTimeout(this.timeout)
      this.timeout = setTimeout(() => {
        this.copied = false
      }, 550)
    }
  }
}
</script>
