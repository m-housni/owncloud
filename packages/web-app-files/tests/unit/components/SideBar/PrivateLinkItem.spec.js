import Vuex from 'vuex'
import GetTextPlugin from 'vue-gettext'
import DesignSystem from '@ownclouders/design-system'
import { mount, createLocalVue } from '@vue/test-utils'
import PrivateLinkItem from 'web-app-files/src/components/SideBar/PrivateLinkItem.vue'

jest.useFakeTimers()

const localVue = createLocalVue()

localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const mapActions = {
  showMessage: jest.fn()
}

describe('PrivateLinkItem', () => {
  it('should render a button', () => {
    const store = createStore()
    const wrapper = getWrapper(store)

    expect(wrapper).toMatchSnapshot()
  })
  it('upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second', async () => {
    const spyShowMessage = jest.spyOn(mapActions, 'showMessage')
    jest.spyOn(window, 'prompt').mockImplementation()

    const store = createStore()
    const wrapper = getWrapper(store)

    expect(spyShowMessage).not.toHaveBeenCalled()

    await wrapper.trigger('click')

    expect(wrapper).toMatchSnapshot()

    expect(spyShowMessage).toHaveBeenCalledTimes(1)

    jest.advanceTimersByTime(550)

    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})

const getTestFolder = () => ({
  type: 'folder',
  ownerId: 'marie',
  ownerDisplayName: 'Marie',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  size: '740',
  name: 'lorem.txt',
  privateLink: 'https://example.com/fake-private-link'
})

function createStore() {
  return new Vuex.Store({
    actions: mapActions,
    commit: jest.fn(),
    dispatch: jest.fn(),
    getters: {
      capabilities: function () {
        return {
          files: {
            privateLinks: true
          }
        }
      }
    },
    modules: {
      Files: {
        namespaced: true,
        getters: {
          highlightedFile: () => getTestFolder()
        }
      }
    }
  })
}

function getWrapper(store) {
  return mount(PrivateLinkItem, {
    localVue,
    store,
    provide: {
      displayedItem: {
        value: getTestFolder()
      }
    }
  })
}
