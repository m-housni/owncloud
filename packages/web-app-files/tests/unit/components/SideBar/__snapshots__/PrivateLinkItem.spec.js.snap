// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateLinkItem should render a button 1`] = `<button aria-label="Copy private link to clipboard" type="button" class="oc-files-private-link-copy-url oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><span>Private link</span> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>`;

exports[`PrivateLinkItem upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second 1`] = `<button aria-label="Copy private link to clipboard" type="button" class="oc-files-private-link-copy-url oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-success oc-button-success-raw"><span>Private link</span> <span class="_clipboard-success-animation oc-icon oc-icon-m oc-icon-passive"><!----></span></button>`;

exports[`PrivateLinkItem upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second 2`] = `<button aria-label="Copy private link to clipboard" type="button" class="oc-files-private-link-copy-url oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><span>Private link</span> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>`;
